acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt

```diff
@@ -0,0 +1,167 @@
+
+    val topic = "stripe-event"
+
+    fun waitForSubscriptionPaymentToSucceed(topic: String = StripeEventListener.topic,
+                                            subscription: String,
+                                            customerId: String,
+                                            timeout: Long = 10000L): Boolean =
+            waitForStripeEvent(
+                    topic = topic,
+                    subscription = subscription,
+                    eventType = "invoice.payment_succeeded",
+                    timeout = timeout,
+                    cmp = {
+                        if (it.type != "invoice.payment_succeeded")
+                            false
+                        else {
+                            val event = it.data.`object` as Invoice
+                            (event.customer == customerId)
+                                    .also {
+                                        if (it)
+                                            logger.info("Received 'invoice.payment_succeeded' event for customer $customerId")
+                                    }
+                        }
+                    })
+
+    fun waitForFailedSubscriptionRenewal(topic: String = StripeEventListener.topic,
+                                         subscription: String,
+                                         customerId: String,
+                                         timeout: Long = 10000L): Boolean =
+            waitForStripeEvent(
+                    topic = topic,
+                    subscription = subscription,
+                    eventType = "customer.subscription.updated",
+                    timeout = timeout,
+                    cmp = {
+                        if (it.type != "customer.subscription.updated")
+                            false
+                        else {
+                            val event = it.data.`object` as Subscription
+                            (event.status == "past_due" && event.customer == customerId)
+                                    .also {
+                                        if (it)
+                                            logger.info("Received 'customer.subscription.updated' from Stripe for customer $customerId with status 'past_due'")
+                                    }
+                        }
+                    })
+
+    private fun waitForStripeEvent(topic: String, subscription: String, eventType: String, timeout: Long, cmp: (Event) -> Boolean): Boolean {
+        var status = false
+        val subscriber = PubSubSubscriber(
+                topic = topic,
+                subscription = subscription,
+                cmp = cmp
+        )
+
+        try {
+            runBlocking {
+                status = withTimeoutOrNull(timeout) {
+                    subscriber.start()
+                    while (!subscriber.eventSeen()) {
+                        delay(200L)
+                    }
+                    "Done"
+                } != null
```

Comment:
This makes it very obscure what we're checking for.   Please consider making a local function with a descriptive name.  Code should preferably be trivial to read, and this isn't.

Reply:
Basically cut and paste from [https://kotlinlang.org/docs/reference/coroutines/cancellation-and-timeouts.html](https://kotlinlang.org/docs/reference/coroutines/cancellation-and-timeouts.html) at the bottom of the page.